### DISCUSSION TITLE
Using composer for dependencies

### DISCUSSION THREAD

**Discussion by yguedidi** [FreshRSS/FreshRSS#4317] [General]
Hello!
I see the codebase contains the code of some external dependencies, why not using Composer instead to install them?
I'm thinking about CssXPath, PPHPMailer and SimplePie
Thanks

**Comment by Frenzie** [comment]
I would suggest giving arguments in favor when proposing something. I.e., how would life improve by using Composer to manage two (!) libraries?

Pretty much the only valid answer is that it would make it easier to add another dozen libraries in the future. For just a handful of libraries it's added complexity, unless the one library you're depending on has a million dependencies.

PS SimplePie is custom, at least in part. Most patches have been accepted upstream.

**Reply by yguedidi** [reply]
Thank you @Frenzie for the fast answer!

Here the advantages I see to use composer, and the use of dependencies:

- it manage the autoloading for us, facilitate adding code and delegate from custom autoloading logic
- rely on a market proven autoloading solution
- allow adding other dependencies very easily, if needed
- relying on dependencies means less code to maintain in the code base
- relying on dependencies avoid home made code that can be less secure, less robust, outdated (exemple: use [Symfony Polyfill](https://github.com/symfony/polyfill) instead of home made polyfills)
- relying on dependencies allow to focus development effort to what provides value, not re-inventing the wheel
- if the dependency needs to be fixed it forces to contribute upstream and collaborate with maintainers, enforcing opensource mindset :slightly_smiling_face: 

> PS SimplePie is custom, at least in part. Most patches have been accepted upstream.

I wanted to play a bit with this so I already managed to have a working prototype locally, and when trying to migrate SimplePie to be a Composer dependency I indeed realized the it's custom :sweat_smile:

**Reply by yguedidi** [reply]
just opened #4319 so you can see the PoC

**Reply by Alkarex** [reply]
Hello @yguedidi and thanks for the suggestion.
For now, the short version is no for production, but yes for development (we [already use it](https://github.com/FreshRSS/FreshRSS/blob/edge/composer.json)) :-)

I do not like depending on random remote services (which might or might not be available, might be hacked), more code, and more libraries for our production service.
All production dependencies should be kept to a minimum, reviewed, and committed along our main codebase. I do not want to promote the approach seen in too many projects of blindly including libraries (not specific to PHP: seems especially bad in Node.js projects, but I see the same in Python, Java...).
Furthermore, some of our dependencies are patched (e.g., SimplePie though with [many contributions upstream](https://github.com/simplepie/simplepie/pulls?q=is%3Apr+is%3Aclosed+author%3AAlkarex)).

The code of `autoload_real.php` is considerably more complicated, longer, and slower than our own, which means more complicated to maintain (for me at least) over the years.

A bit related: https://github.com/FreshRSS/FreshRSS/issues/4129

**Reply by yguedidi** [reply]
Thank you for the detailed answer @Alkarex.

> All production dependencies should be kept to a minimum, reviewed, and committed along our main codebase.

I completely understand your concerns!

Especially if we talk about the production ZIP package that people just have to extract on their server :slightly_smiling_face: 
Seems that from your point of view this is the production asset, and I'm fine with that.
But do this mean that the Git repository for development purpose must be that production asset? I don't think so.

Actually the final ZIP package looks to me more like the result of a build process, from source code.
And source code should be developer friendly, to facilitate development, and onboarding of new contributors.
Composer is the de facto autoloading and dependency tool now in PHP world.
Having those things managed by Composer mean 2 less things to be managed by contributors, focusing on what provides value: features!
And the setup will be more modern!

Going back to the build process idea, with Composer it could be that we run `composer install --optimize-autoloader` and then build a ZIP archive including vendor folder.
Composer install vendors from distribution, this means that it downloads the ZIP archive of each vendor from GitHub.
Many libraries are using a `.gitattributes` file to control which files are includes in the ZIP, this allow excluding docs, tests, useless files etc to keep only production code in the ZIP, so the library is lightweight.
And the optimized autoloader is a simple array map from class FQCN to file path, so quite sure it will perform at least as good as the custom autoloading, if not better :wink:

> I do not want to promote the approach seen in too many projects of blindly including libraries

Blindly including libraries is a developer issue, it's not the tool to be blamed haha!
Dependencies must always be carefully chosen!

> Furthermore, some of our dependencies are patched (e.g., SimplePie though with [many contributions upstream](https://github.com/simplepie/simplepie/pulls?q=is%3Apr+is%3Aclosed+author%3AAlkarex)).

During development it's possible to depend on your fork meanwhile fixes are merged and released upstream.

> The code of autoload_real.php is considerably more complicated, longer, and slower than our own, **which means more complicated to maintain** (for me at least) over the years.

It's a generated file, so no maintenance at all from you and all contributors

**Reply by yguedidi** [reply]
So in the end, best of both worlds:
- developer friendly repository, relying on standard Composer, delegating autoloading and dependencies, focusing on feature
- production ready build asset, including vendor, in their lightweight versions

**Reply by Alkarex** [reply]
This is how `autoload_real.php` looks for me (but maybe I have missed another option), which is not especially nice:

```php
<?php

// autoload_real.php @generated by Composer

class ComposerAutoloaderInitbcf427b8ec561e12cda3fa233fb1a6dc
{
    private static $loader;

    public static function loadClassLoader($class)
    {
        if ('Composer\Autoload\ClassLoader' === $class) {
            require __DIR__ . '/ClassLoader.php';
        }
    }

    /**
     * @return \Composer\Autoload\ClassLoader
     */
    public static function getLoader()
    {
        if (null !== self::$loader) {
            return self::$loader;
        }

        require __DIR__ . '/platform_check.php';

        spl_autoload_register(array('ComposerAutoloaderInitbcf427b8ec561e12cda3fa233fb1a6dc', 'loadClassLoader'), true, true);
        self::$loader = $loader = new \Composer\Autoload\ClassLoader(\dirname(\dirname(__FILE__)));
        spl_autoload_unregister(array('ComposerAutoloaderInitbcf427b8ec561e12cda3fa233fb1a6dc', 'loadClassLoader'));

        $useStaticLoader = PHP_VERSION_ID >= 50600 && !defined('HHVM_VERSION') && (!function_exists('zend_loader_file_encoded') || !zend_loader_file_encoded());
        if ($useStaticLoader) {
            require __DIR__ . '/autoload_static.php';

            call_user_func(\Composer\Autoload\ComposerStaticInitbcf427b8ec561e12cda3fa233fb1a6dc::getInitializer($loader));
        } else {
            $map = require __DIR__ . '/autoload_namespaces.php';
            foreach ($map as $namespace => $path) {
                $loader->set($namespace, $path);
            }

            $map = require __DIR__ . '/autoload_psr4.php';
            foreach ($map as $namespace => $path) {
                $loader->setPsr4($namespace, $path);
            }

            $classMap = require __DIR__ . '/autoload_classmap.php';
            if ($classMap) {
                $loader->addClassMap($classMap);
            }
        }

        $loader->register(true);

        if ($useStaticLoader) {
            $includeFiles = Composer\Autoload\ComposerStaticInitbcf427b8ec561e12cda3fa233fb1a6dc::$files;
        } else {
            $includeFiles = require __DIR__ . '/autoload_files.php';
        }
        foreach ($includeFiles as $fileIdentifier => $file) {
            composerRequirebcf427b8ec561e12cda3fa233fb1a6dc($fileIdentifier, $file);
        }

        return $loader;
    }
}

function composerRequirebcf427b8ec561e12cda3fa233fb1a6dc($fileIdentifier, $file)
{
    if (empty($GLOBALS['__composer_autoload_files'][$fileIdentifier])) {
        require $file;

        $GLOBALS['__composer_autoload_files'][$fileIdentifier] = true;
    }
}
```

**Reply by yguedidi** [reply]
@Alkarex if I understand correctly, you don't trust Composer for autoloading?

**Reply by yguedidi** [reply]
It's a perfectly valid position, just asking to know if there is any chance that Composer get used, or I should abandon arguing :sweat_smile:

**Reply by Frenzie** [reply]
> And source code should be developer friendly, to facilitate development, and onboarding of new contributors.

I happen to disagree that adding Composer would do that. :-) It adds a complicated to set up additional step for each individual developer, which is a necessary evil once you get past a certain point. But it's gatekeeping, plain and simple. One might of course take the position that this is desirable as a kind of filter against people who are too inexperienced. Conversely, I don't see how Composer would *attract* anyone.

**Reply by yguedidi** [reply]
Hello @Frenzie , thank you for your feedback sharing!
Me I tend to not focus on the tool itself (from my perspective it's a matter of few commands to set up, and `composer install`), but more on that architecture: rely on proven, tested, dedicated 3rd party libraries, not reinventing the wheel
How this help new contributors? less code to learn! less code to adapt to! only feature code!

**Reply by Frenzie** [reply]
From mine and @Alkarex' perspective too! Even if Composer were completely unknown to us, it'd be only one tiny thing we didn't know.

> but more on that architecture: rely on proven, tested, dedicated 3rd party libraries, not reinventing the wheel

We do all that already, though. ;-)

**Reply by Alkarex** [reply]
> @Alkarex if I understand correctly, you don't trust Composer for autoloading?

Not so much *not trusting* but more *not liking* (for production in the context of this specific project), both from readability / maintainability and performance view points. Maybe if that part improves in the future :-)

**Reply by yguedidi** [reply]
> We do all that already, though. ;-)

True! :sweat_smile: 

So the discussion should be rebased on using a dependency manager or not I think :)
Pros:
- lighter repository, only our code
- dependency config file serve as documentation, listing all external dependencies
- facilitate adding new dependencies, deals with the autoload
- facilitate updating dependencies

**Reply by yguedidi** [reply]
> both from readability / maintainability and performance view points

@Alkarex could you be more precise here please? are you talking about Composer itself?

**Reply by Alkarex** [reply]
@yguedidi If you look at the code of the autoloader I posted a bit higher up, it is really not something I would like to use in our case, with loops, call_user_func, maps, more classes, more requires... That is what I mean by the readability / maintainability and performance regressions that would bring.

**Reply by ColonelMoutarde** [reply]
> > And source code should be developer friendly, to facilitate development, and onboarding of new contributors.
> 
> I happen to disagree that adding Composer would do that. :-) It adds a complicated to set up additional step for each individual developer, which is a necessary evil once you get past a certain point. But it's gatekeeping, plain and simple. One might of course take the position that this is desirable as a kind of filter against people who are too inexperienced. Conversely, I don't see how Composer would _attract_ anyone.

Hi, for me composer it's, the first step in modern PHP application :)

**Reply by yguedidi** [reply]
@Alkarex thank you for the details!

> it is really not something I would like to use in our case

I respect this. It seems to me that you consider vendor code as code you own like your project code.
It's not my position, as soon as the vendor is trusted and verified.
maybe that's why we will not agree for this project :)

I will open a more complete discussion about modernizing FreshRSS, with my vision, to get your feedback

**Reply by Art4** [reply]
@Frenzie 

> I would suggest giving arguments in favor when proposing something. I.e., how would life improve by using Composer to manage two (!) libraries?

Sure. Let's say you are working on a library that comes with its own cache implementation. This cache implementation is actually out of scope of your library and you would like to delete your own implementation. But you still want to allow your users to put their own cache implementation into your library.

Fortunately, there is a standard recommendation for this: [PSR-16](https://www.php-fig.org/psr/psr-16/). You can simply install it with Composer: `php composer.phar require psr/simple-cache` and then program against the interfaces in your library. Next you can simply delete your cache implementation (or implement a simple fake cache), which disables caching by default. If your users want the library to use a cache, then you simply tell them to inject a PSR-16 implementation in your library.

With this you have created **only 1** (!) Composer dependency in your library. In fact, in this example you only have a dependency on interfaces, not classes. But with this you have improved life for everyone:
1. the user can use his own cache implementation. 
2. you as library maintainer can delete code and have less code to maintain.
3. the user can change the cache implementation at any time if something is wrong with it. You as library maintainer are absolutely not affected anymore.

**Reply by Frenzie** [reply]
I'm not convinced any of that is even tangentially relevant here, but that's an interesting example nonetheless. ^_^

**Reply by Art4** [reply]
@Frenzie 

> I'm not convinced any of that is even tangentially relevant here, but that's an interesting example nonetheless. ^_^

Thank you. Then I would like to be more specific about how this example will directly affect FreshRSS in the future. I came across this discussion because I am a contributor to SimplePie and want to understand more how SimplePie is used in projects. 

With SimplePie 1.6, there is initial support for namespaces and PSR-4, see [#711](https://github.com/simplepie/simplepie/pull/711). The next step will be to deprecate the use of SimplePie without namespaces (because PSR-0 is deprecated too), see [#722](https://github.com/simplepie/simplepie/pull/722).

After that, it will be possible to write the cache implementation against PSR-16 interfaces, as explained in my example. And also the HTTP requests can run over an HTTP library thanks to PSR-7, see [#520](https://github.com/simplepie/simplepie/issues/520#issuecomment-1046666308). But for this it will be necessary to have the PSR interfaces in the code. Instead of setting them hard in the code, they will be installed via Composer.

In FreshRSS, a fork of SimplePie is used, incorporating the major changes from SimplePie. This will become more complicated in the future because of the development described above and will require more work.

Like here at FreshRSS, I used to manage dependencies in my projects without Composer for years until I understood Composer deeply enough. I could have saved myself years of work by going straight to Composer. Still, it was worth that work to me because I had to dig deep into Composer and I learned a lot. 

So I understand that you are skeptical about Composer at first. Your point that it is easy to accumulate way too many dependencies, I can absolutely understand. But that is not a problem or the responsibility of Composer, but of the developer. And in my example I showed how a wisely taken dependency can lead to less, rather than more dependencies.

I don't want to say that Composer is always and everywhere the best solution, but especially for FreshRSS I see Composer as the right tool. So I would strongly encourage you to look deeper into Composer as a dependency manager as soon as possible. Using it in the dev environment like @Alkarex said is an important first step. :+1:

**Reply by Frenzie** [reply]
I already covered all off the above in "For just a handful of libraries it's added complexity, unless the one library you're depending on has a million dependencies." and "which is a necessary evil once you get past a certain point." ;-)

> Like here at FreshRSS, I used to manage dependencies in my projects without Composer for years until I understood Composer deeply enough. I could have saved myself years of work by going straight to Composer. Still, it was worth that work to me because I had to dig deep into Composer and I learned a lot.

We all use Composer, PyPI, LuaRocks, NPM, Gradle, etc. It's not an unknown entity.
